File: frontend/src/shared/api/regions.ts
/**
 * Regions API client
 */

import { apiClient } from './client'

// Types
export interface RegionSummary {
  region_id: number
  region_name: string
  status: string
  total_players: number
  active_players: number
  is_active: boolean
  created_at: string
}

export interface RegionListResponse {
  regions: RegionSummary[]
  total?: number
}

// API functions
export const regionsApi = {
  /**
   * Get all regions (public endpoint for registration)
   */
  async getRegions(activeOnly: boolean = false): Promise<RegionListResponse> {
    const startTime = Date.now()
    console.log('=== 发起赛区数据请求 ===', {
      activeOnly,
      timestamp: new Date().toISOString(),
    })

    try {
      const params = new URLSearchParams()
      if (activeOnly) {
        params.append('active_only', 'true')
      }

      const queryString = params.toString()
      const url = queryString ? `/registration/regions?${queryString}` : '/registration/regions'

      console.log('请求URL:', url)

      const response = await apiClient.get(url)
      const duration = Date.now() - startTime

      console.log('=== 赛区数据请求成功 ===', {
        url,
        duration: `${duration}ms`,
        regionsCount: response.data.regions?.length || 0,
        regions:
          response.data.regions?.map((r: any) => ({ id: r.region_id, name: r.region_name })) || [],
      })

      return response.data
    } catch (error: any) {
      const duration = Date.now() - startTime
      console.error('=== 赛区数据请求失败 ===', {
        duration: `${duration}ms`,
        error: {
          status: error.response?.status,
          message: error.message,
          code: error.code,
        },
      })

      // 不再使用模拟数据，直接抛出错误让调用方处理
      console.error('Regions API 请求失败，无法获取真实数据')
      throw error
    }
  },

  /**
   * Get all regions (authenticated endpoint)
   */
  async getRegionsAuth(activeOnly: boolean = false): Promise<RegionListResponse> {
    const params = new URLSearchParams()
    if (activeOnly) {
      params.append('active_only', 'true')
    }

    const queryString = params.toString()
    const url = queryString ? `/regions?${queryString}` : '/regions'

    const response = await apiClient.get(url)
    return response.data
  },

  /**
   * Get region by ID
   */
  async getRegion(regionId: string): Promise<RegionSummary> {
    const response = await apiClient.get(`/regions/${regionId}`)
    return response.data
  },
}

File: backend/src/presentation/api/routes/regions.py
"""
Regions API routes.
"""

from fastapi import APIRouter, Depends, HTTPException, status
from typing import List
import structlog

from src.presentation.schemas.player import RegionSummaryResponse, RegionListResponse
from src.application.services.player_service import PlayerService
from src.presentation.dependencies.player import get_player_service
from src.presentation.dependencies.permission import verify_permission
from src.domain.value_objects.role import PermissionAction, PermissionResource


router = APIRouter()
logger = structlog.get_logger(__name__)


@router.get("/health")
async def regions_health():
    """Regions service health check"""
    return {"status": "Regions service is healthy"}


@router.get("/public", response_model=RegionListResponse)
async def get_regions_public(active_only: bool = False):
    """
    Get all regions - public endpoint for registration.

    This endpoint provides access to regions without authentication,
    specifically for use during player registration and public viewing.

    Args:
        active_only: If True, only returns active regions
    """
    try:
        from src.infrastructure.database.connection import database_manager
        from src.infrastructure.repositories.region import SQLAlchemyRegionRepository
        
        async with database_manager.get_session() as session:
            region_repository = SQLAlchemyRegionRepository(session)
            
            if active_only:
                regions = await region_repository.find_all_active()
            else:
                regions = await region_repository.find_all()

            return RegionListResponse(
                regions=[
                    RegionSummaryResponse(
                        region_id=region.region_id,
                        region_name=region.region_name,
                        status=region.status,
                        total_players=region.total_players,
                        active_players=region.active_players,
                        is_active=region.is_active,
                        created_at=region.created_at,
                    )
                    for region in regions
                ],
                total=len(regions),
            )

    except Exception as e:
        logger.error(
            "Failed to get regions (public)",
            active_only=active_only,
            error=str(e),
            error_type=type(e).__name__,
        )
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR, detail="获取赛区列表失败，请稍后重试"
        )


@router.get("/", response_model=RegionListResponse)
@router.get("", response_model=RegionListResponse)  # 支持不带斜杠的路径
async def get_regions(
    active_only: bool = False,
    player_service: PlayerService = Depends(get_player_service),
    _: None = Depends(
        verify_permission(PermissionResource.REGION, PermissionAction.READ)
    ),
):
    """
    Get all regions.

    Returns a list of all regions in the system. Can be filtered to show only active regions.

    Args:
        active_only: If True, only returns active regions
    """
    try:
        if active_only:
            regions = await player_service.get_active_regions()
        else:
            regions = await player_service.get_all_regions()

        return RegionListResponse(
            regions=[
                RegionSummaryResponse(
                    region_id=region.region_id,
                    region_name=region.region_name,
                    status=region.status,
                    total_players=region.total_players,
                    active_players=region.active_players,
                    is_active=region.is_active,
                    created_at=region.created_at,
                )
                for region in regions
            ],
            total=len(regions),
        )

    except Exception as e:
        logger.error(
            "Failed to get regions",
            active_only=active_only,
            error=str(e),
            error_type=type(e).__name__,
        )
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR, detail="获取赛区列表失败，请稍后重试"
        )


@router.get("/{region_id}", response_model=RegionSummaryResponse)
async def get_region(
    region_id: str,
    player_service: PlayerService = Depends(get_player_service),
    _: None = Depends(
        verify_permission(PermissionResource.REGION, PermissionAction.READ)
    ),
):
    """
    Get a specific region by ID.

    Returns detailed information about a region.
    """
    try:
        region = await player_service.get_region_by_id(region_id)

        if not region:
            raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="赛区未找到")

        return RegionSummaryResponse(
            region_id=region.region_id,
            region_name=region.region_name,
            status=region.status,
            total_players=region.total_players,
            active_players=region.active_players,
            is_active=region.is_active,
            created_at=region.created_at,
        )

    except HTTPException:
        raise
    except Exception as e:
        logger.error(
            "Failed to get region",
            region_id=region_id,
            error=str(e),
            error_type=type(e).__name__,
        )
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR, detail="获取赛区信息失败，请稍后重试"
        )
